File: solutions/python/04/abc177_d/main.py
import sys


class DSU:
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.size = [1] * n

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    def unite(self, a: int, b: int) -> None:
        a, b = self.find(a), self.find(b)
        if a == b:
            return
        if self.size[a] < self.size[b]:
            a, b = b, a
        self.parent[b] = a
        self.size[a] += self.size[b]

    def same(self, a: int, b: int) -> bool:
        return self.find(a) == self.find(b)


def main() -> None:
    input = sys.stdin.readline
    n, q = map(int, input().split())
    dsu = DSU(n)
    out = []
    for _ in range(q):
        t, u, v = map(int, input().split())
        u -= 1
        v -= 1
        if t == 1:
            dsu.unite(u, v)
        else:
            out.append("Yes" if dsu.same(u, v) else "No")
    print("\n".join(out))


if __name__ == "__main__":
    main()

File: solutions/go/04/abc177_d/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type DSU struct {
	parent, size []int
}

func NewDSU(n int) *DSU {
	p, s := make([]int, n), make([]int, n)
	for i := range p {
		p[i] = i
		s[i] = 1
	}
	return &DSU{p, s}
}

func (d *DSU) Find(x int) int {
	for d.parent[x] != x {
		d.parent[x] = d.parent[d.parent[x]]
		x = d.parent[x]
	}
	return x
}

func (d *DSU) Unite(a, b int) {
	a, b = d.Find(a), d.Find(b)
	if a == b {
		return
	}
	if d.size[a] < d.size[b] {
		a, b = b, a
	}
	d.parent[b] = a
	d.size[a] += d.size[b]
}

func (d *DSU) Same(a, b int) bool {
	return d.Find(a) == d.Find(b)
}

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)
	d := NewDSU(n)
	for ; q > 0; q-- {
		var t, u, v int
		fmt.Fscan(in, &t, &u, &v)
		u--
		v--
		if t == 1 {
			d.Unite(u, v)
		} else {
			if d.Same(u, v) {
				fmt.Fprintln(out, "Yes")
			} else {
				fmt.Fprintln(out, "No")
			}
		}
	}
}
